server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java

```diff
@@ -64,9 +100,93 @@ public String getName() {
 
     @Override
     public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
-        ClusterStatsRequest clusterStatsRequest = new ClusterStatsRequest().nodesIds(request.paramAsStringArray("nodeId", null));
+        Set<String> metrics = Strings.tokenizeByCommaToSet(request.param("metric", "_all"));
+        Set<String> indexMetrics = Strings.tokenizeByCommaToSet(request.param("index_metric", null));
+        String[] nodeIds = request.paramAsStringArray("nodeId", null);
+
+        ClusterStatsRequest clusterStatsRequest = new ClusterStatsRequest().nodesIds(nodeIds);
         clusterStatsRequest.timeout(request.param("timeout"));
         clusterStatsRequest.useAggregatedNodeLevelResponses(true);
+
+        if (!metrics.isEmpty()) {
+            if (metrics.size() > 1 && metrics.contains("_all")) {
+                throw new IllegalArgumentException(
+                    String.format(
+                        Locale.ROOT,
+                        "request [%s] contains _all and individual metrics [%s]",
+                        request.path(),
+                        request.param("metric")
+                    )
+                );
+            }
+
+            if (indexMetrics.size() > 1 && indexMetrics.contains("_all")) {
+                throw new IllegalArgumentException(
+                    String.format(
+                        Locale.ROOT,
+                        "request [%s] contains _all and individual index metrics [%s]",
+                        request.path(),
+                        request.param("index_metric")
+                    )
+                );
+            }
+
+            clusterStatsRequest.applyMetricFiltering(true);
+
+            final Set<String> metricsRequested = new HashSet<>();
+            if (metrics.contains("_all")) {
+                metricsRequested.addAll(METRIC_REQUEST_CONSUMER_MAP.keySet());
```

Comment:
should we add `indexMetricsRequested.addAll(INDEX_METRIC_TO_REQUEST_CONSUMER_MAP.keySet());` as well here ?

Reply:
Yes populate the subset in that case.